### PR TITLE
Adds measure and sample for density matrices.

### PR DESCRIPTION
**Pull request by dabacon**

### CODE REVIEW

File: cirq/sim/density_matrix.py [hunk 1]
```diff
@@ -71,3 +72,180 @@ def to_valid_density_matrix(
...
+    # Shape into a tensor
+    tensor = np.reshape(all_probs, [2] * num_qubits)
+
+    # Calculate the probabilities for measuring the particular results.
+    probs = [
+        np.sum(np.abs(tensor[linalg.slice_for_qubits_equal_to(indices, b)]))
+        for b in range(2 ** len(indices))]
+
+    # To deal with rounding issues, ensure that the probabilities sum to 1.
+    probs /= sum(probs) # type: ignore
```
Comment (Strilanc) [inline cirq/sim/density_matrix.py]: `np.sum` will be faster

File: cirq/sim/density_matrix.py [hunk 2]
```diff
@@ -71,3 +72,180 @@ def to_valid_density_matrix(
...
+    return probs
+
+
+def _validate_num_qubits(matrix: np.ndarray) -> int:
+    """Validates that matrix's shape is a valid shape for qubits.
+    """
+    shape = matrix.shape
+    half_index = len(shape) // 2
```
Comment (Strilanc) [inline cirq/sim/density_matrix.py]: Did you check that the length is even anywhere?

File: cirq/sim/density_matrix.py [hunk 3]
```diff
@@ -71,3 +72,180 @@ def to_valid_density_matrix(
...
+    row_size = np.prod(shape[:half_index]) if shape != (0,) else 0
```
Comment (Strilanc) [inline cirq/sim/density_matrix.py]: Why would the shape ever be `(0,)`? It should be `()` instead in the no-qubit case.

File: cirq/sim/density_matrix.py [hunk 4]
```diff
@@ -71,3 +72,180 @@ def to_valid_density_matrix(
...
+    col_size = np.prod(shape[half_index:]) if shape != (0,) else 0
+    if row_size != col_size:
+        raise ValueError(
+            'Matrix was not square. Shape was {}'.format(shape))
+    if row_size !=0 and row_size & (row_size - 1):
```
Comment (Strilanc) [inline cirq/sim/density_matrix.py]: `row_size & (row_size - 1)` being true implies `row_size != 0`.

File: cirq/sim/density_matrix.py [hunk 5]
```diff
@@ -71,3 +72,180 @@ def to_valid_density_matrix(
...
+        ValueError if the dimension of the matrix is not compatible with a
+            matrix of n qubits.
+        IndexError if the indices are out of range for the number of qubits
+            corresponding to the density matrix.
+    """
+    num_qubits = _validate_num_qubits(matrix)
+    _validate_indices(num_qubits, indices)
+
+    if len(indices) == 0:
+        return ([], np.copy(matrix))
```
[the remark this answers is not in the document]
Comment (dabacon) [reply cirq/sim/density_matrix.py]: Added test.

File: cirq/sim/density_matrix.py (hunk 1 above)
Comment (dabacon) [reply cirq/sim/density_matrix.py]: done

**Comment by dabacon** [comment]
Ping on this one.

**Pull request merged**